### PR TITLE
[JUJU-1684] Commit secret updates, deletes, access changes at end of hook execution

### PR DESCRIPTION
**Pull request by wallyworld**
Secret updates, deletes, access changes are cached on the hook context and committed when the hook has finished running.
Creates still need to happen immediately since the returned URI is needed for subsequent operations.

Also fix some unit agent context API calls to use a secret uri not a string. 

## Checklist

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```
juju deploy juju-qa-dummy-source
Located charm "juju-qa-dummy-source" in charm-hub, revision 6
Deploying "juju-qa-dummy-source" from charm-hub charm "juju-qa-dummy-source", revision 6 in channel stable on jammy
juju deploy juju-qa-dummy-sink -n 2
Located charm "juju-qa-dummy-sink" in charm-hub, revision 7
Deploying "juju-qa-dummy-sink" from charm-hub charm "juju-qa-dummy-sink", revision 7 in channel stable on jammy

juju exec --unit dummy-source/0 "secret-add foo=bar --label=baz"
secret:cc1h4l2ov0ni3c50g0qg
juju exec --unit dummy-source/0 "secret-ids"
secret:cc1h4l2ov0ni3c50g0qg: baz
juju exec --unit dummy-source/0 "secret-get secret:cc1h4l2ov0ni3c50g0qg"
foo: bar

juju exec --unit dummy-source/0 "secret-update secret:cc1h4l2ov0ni3c50g0qg foo=bar3"
juju exec --unit dummy-source/0 "secret-get secret:cc1h4l2ov0ni3c50g0qg"
foo: bar
juju exec --unit dummy-source/0 "secret-get secret:cc1h4l2ov0ni3c50g0qg --update"
foo: bar3

juju relate dummy-sink dummy-source
juju exec --unit dummy-source/0 "secret-grant secret:cc1h4l2ov0ni3c50g0qg -r 0 --unit dummy-sink/0"
juju exec --unit dummy-sink/0 "secret-get secret:cc1h4l2ov0ni3c50g0qg"
foo: bar3

juju exec --unit dummy-source/0 "secret-revoke secret:cc1h4l2ov0ni3c50g0qg -r 0 --unit dummy-sink/0"
juju exec --unit dummy-sink/0 "secret-get secret:cc1h4l2ov0ni3c50g0qg"
ERROR permission denied

```

### CODE REVIEW

File: apiserver/facades/agent/uniter/uniter.go [hunk 1]
```diff
@@ -2718,6 +2720,43 @@ func (u *UniterAPI) commitHookChangesForOneUnit(unitTag names.UnitTag, changes p
 		modelOps = append(modelOps, modelOp)
 	}
 
+	if len(changes.SecretDeletes) > 0 {
+		result, err := u.SecretsManagerAPI.RemoveSecrets(params.SecretURIArgs{Args: changes.SecretDeletes})
```
Comment (hpidcock) [inline apiserver/facades/agent/uniter/uniter.go:2724]: obviously would be great to get these into the txn

File: worker/uniter/runner/context/contextfactory.go [hunk 2]
```diff
@@ -65,12 +64,8 @@ type StorageContextAccessor interface {
 
 type SecretsAccessor interface {
 	SecretIds() (map[*secrets.URI]string, error)
-	Create(cfg *secrets.SecretConfig, ownerTag names.Tag, value secrets.SecretValue) (string, error)
-	Update(uri string, cfg *secrets.SecretConfig, value secrets.SecretValue) error
-	Remove(uri string) error
-	GetValue(uri, label string, update, peek bool) (secrets.SecretValue, error)
-	Grant(uri string, p *secretsmanager.SecretRevokeGrantArgs) error
-	Revoke(uri string, p *secretsmanager.SecretRevokeGrantArgs) error
+	Create(cfg *secrets.SecretConfig, ownerTag names.Tag, value secrets.SecretValue) (*secrets.URI, error)
```
Comment (hpidcock) [inline worker/uniter/runner/context/contextfactory.go:67]: Will Create be moved into the Commit phase of hook execution like we had discussed?

**Review by hpidcock — APPROVED** [review]

File: worker/uniter/runner/context/contextfactory.go (hunk 2 above)
Comment (wallyworld) [reply worker/uniter/runner/context/contextfactory.go:67]: Yes, need to expose the means to get the URI ahead of time first

File: apiserver/facades/agent/uniter/uniter.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/facades/agent/uniter/uniter.go:2724]: yeah. we're getting into 2 phase commit territory if the backend is not juju etc.
i'd like to come back to this when the design has evolved a bit, will leave a todo

**Pull request merged**